packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts

```diff
@@ -591,6 +592,155 @@ async function updateAutoLink(
   }
 }
 
+export function ensureWinUIDialect(
+  slnFile: string,
+  windowsProjectConfig: WindowsProjectConfig,
+  checkMode: boolean,
+) {
+  const buildFlagsProps = path.join(path.dirname(slnFile), 'BuildFlags.props');
+  let changesNeeded = false;
+  if (fs.existsSync(buildFlagsProps)) {
+    const buildFlagsContents = configUtils.readProjectFile(buildFlagsProps);
+    const useWinUI3FromConfig = windowsProjectConfig.useWinUI3;
+    const useWinUI3FromBuildFlags =
+      configUtils
+        .tryFindPropertyValue(buildFlagsContents, 'UseWinUI3')
+        ?.toLowerCase() === 'true';
+    // use the UseWinUI3 value in react-native.config.js, or if not present, the value from BuildFlags.props
+    changesNeeded = fixPackagesConfig(
+      useWinUI3FromConfig !== undefined
+        ? useWinUI3FromConfig
+        : useWinUI3FromBuildFlags,
+      windowsProjectConfig,
+      checkMode,
+    );
+    if (useWinUI3FromConfig !== undefined) {
+      // Make sure BuildFlags matches the value that comes from react-native.config.js
+      const node = buildFlagsContents.getElementsByTagName('UseWinUI3');
+      const newValue = useWinUI3FromConfig ? 'true' : 'false';
+      changesNeeded = changesNeeded || node.item(0)?.textContent !== newValue;
+      if (!checkMode && changesNeeded) {
+        node.item(0)!.textContent = newValue;
+        const buildFlagsOutput = new XMLSerializer().serializeToString(
+          buildFlagsContents,
+        );
+        fs.writeFileSync(buildFlagsProps, buildFlagsOutput);
+      }
+    }
+  }
+  return changesNeeded;
+}
+
+function fixPackagesConfig(
+  useWinUI3: boolean,
+  project: WindowsProjectConfig,
+  checkMode: boolean,
+) {
+  const projectFile = path.join(
+    project.folder,
+    project.sourceDir,
+    project.project!.projectFile,
+  );
+  const packagesConfig = path.join(
+    path.dirname(projectFile),
+    'packages.config',
+  );
+
+  let changed = false;
+
+  if (fs.existsSync(packagesConfig)) {
+    // if we don't have a packages.config, then this is a C# project, in which case we use <PackageReference> and dynamically pick the right XAML package.
+    const packagesConfigContents = configUtils.readProjectFile(packagesConfig);
+    const packageElements = packagesConfigContents.documentElement.getElementsByTagName(
+      'package',
+    );
+    const winUIPropsPath = path.join(
+      resolveRnwRoot(project),
+      'PropertySheets/WinUI.props',
+    );
+    const winuiPropsContents = configUtils.readProjectFile(winUIPropsPath);
+    const winui2xVersion = configUtils.tryFindPropertyValue(
+      winuiPropsContents,
+      'WinUI2xVersion',
+    );
+    const winui3Version = configUtils.tryFindPropertyValue(
+      winuiPropsContents,
+      'WinUI3Version',
+    );
+
+    const dialects = [
+      {id: 'Microsoft.WinUI', version: winui3Version!},
+      {id: 'Microsoft.UI.Xaml', version: winui2xVersion!},
+    ];
+    const keepPkg = useWinUI3 ? dialects[0] : dialects[1];
```

Comment:
Is there any way to extract out steps for adding/updating/removing nodes in a packages.config into their own utility functions? Then I could re-use it when autolinking needs to do the same.

Reply:
i'll take a look in the refactor PR :)